src/java.desktop/share/classes/javax/swing/ImageIcon.java

```diff
@@ -436,10 +439,28 @@ public void setDescription(String description) {
      * @param y the Y coordinate of the icon's top-left corner
      */
     public synchronized void paintIcon(Component c, Graphics g, int x, int y) {
-        if(imageObserver == null) {
-           g.drawImage(image, x, y, c);
-        } else {
-           g.drawImage(image, x, y, imageObserver);
+        Object oldHint = null;
+        boolean hintChanged = false;
+        ImageObserver observer = imageObserver != null ? imageObserver : c;
+        if (image instanceof MultiResolutionImage) {
+            Image variant = ((MultiResolutionImage) image).getResolutionVariant(this.width, this.height);
+            if (variant.getHeight(observer) != height || variant.getWidth(observer) != width) {
+                if (g instanceof Graphics2D) {
+                    Graphics2D g2d = (Graphics2D) g;
+                    RenderingHints rh = g2d.getRenderingHints();
+                    oldHint = rh.get(RenderingHints.KEY_INTERPOLATION);
+                    if (!RenderingHints.VALUE_INTERPOLATION_BICUBIC.equals(oldHint)) {
+                        g2d.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
+                                RenderingHints.VALUE_INTERPOLATION_BICUBIC);
+                        hintChanged = true;
+                    }
+                }
+            }
+        }
+        g.drawImage(image, x, y, observer);
+        if (hintChanged) {
+            ((Graphics2D) g).setRenderingHint(RenderingHints.KEY_INTERPOLATION,
+                    oldHint == null ? RenderingHints.VALUE_INTERPOLATION_NEAREST_NEIGHBOR : oldHint);
```

Comment:
This might be a rookie question:

When it comes to modifying and restoring a Graphics, I've seen two common patterns:
A. Call `newG = g.create()` and later `newG.dispose()`
B. Call `g.setX(newValue)` and then call `g.setX(oldValue)` later

Is there guidance on when to use which? Or is one always preferred?

Reply:
> This might be a rookie question:
> 
> When it comes to modifying and restoring a Graphics, I've seen two common patterns: A. Call `newG = g.create()` and later `newG.dispose()` B. Call `g.setX(newValue)` and then call `g.setX(oldValue)` later
> 
> Is there guidance on when to use which? Or is one always preferred?

Usually i follow the rule of minimal action. When i need to set a lot of hints that needs to be later unrolled such as changing graphics coordinates and such - i'm creating a new Graphics. When i only need to manipulate a single hint and it can easily be unrolled - i am using the same object. In this case i'm changing just a single hint.

Reply:
I am not sure that this is the right place to change the hints.
If we want to always scale the icon using the bicubic interpolation then it is better to do it where we call the paintIcon() method in some jcomponent(this probably may be configured by some option - similar to the text antialisaing). But it will affect performance since the bicubic is much slower.

But before discussing this shared code change could you please provide some info on why it worked fine before? Why we try to downscale the image and do not get an exact size from the native.

Reply:
I suggest to recheck the issues described above, updating the shared implementation of `ImageIcon.java` seems wrong